internal/dataplane/kongstate/kongupstreampolicy.go

```diff
@@ -14,14 +20,56 @@ const (
 	KongHashOnTypeURICapture string = "uri_capture"
 )
 
+// GetKongUpstreamPolicyForServices scans all Services in the group to see if their KongUpstreamPolicy is consistent
+// and returns a non-nil KongUpstreamPolicy if it is.
+//
+// We require either:
+// - all the Services to be configured with the same KongUpstreamPolicy.
+// - none of the Services to be configured with a KongUpstreamPolicy.
+//
+// If the KongUpstreamPolicy configuration is inconsistent or a configured KongUpstreamPolicy cannot be fetched from
+// the store, an error is returned.
+func GetKongUpstreamPolicyForServices(s store.Storer, servicesGroup []*corev1.Service) (*kongv1beta1.KongUpstreamPolicy, error) {
+	if len(servicesGroup) == 0 {
+		return nil, nil
+	}
+
+	servicesGroupedByUpstreamPolicy := lo.GroupBy(servicesGroup, func(svc *corev1.Service) mo.Option[string] {
+		policyName, ok := annotations.ExtractUpstreamPolicy(svc.Annotations)
```

Comment:
Is it guaranteed that at least one service has upstream policy annotation? If not, it is possible to get an empty `servicesGroupedByUpstreamPolicy`.

Reply:
Yes, it is, because we group by `mo.Option[string]`. In the case of `mo.None[string]` (meaning the annotation is absent), it will also create a separate group.